core/config/toml/types.go

```diff
@@ -713,6 +752,110 @@ func (w *WebServerTLS) setFrom(f *WebServerTLS) {
 	}
 }
 
+type WebServerLDAP struct {
+	ServerTLS                   *bool
+	SessionTimeout              *models.Duration
+	QueryTimeout                *models.Duration
+	BaseUserAttr                *string
+	BaseDN                      *string
+	UsersDN                     *string
+	GroupsDN                    *string
+	ActiveAttribute             *string
+	ActiveAttributeAllowedValue *string
+	AdminUserGroupCN            *string
+	EditUserGroupCN             *string
+	RunUserGroupCN              *string
+	ReadUserGroupCN             *string
+	UserApiTokenEnabled         *bool
+	UserAPITokenDuration        *models.Duration
+	UpstreamSyncInterval        *models.Duration
+	UpstreamSyncRateLimit       *models.Duration
+}
+
+func (w *WebServerLDAP) setFrom(f *WebServerLDAP) {
+	if v := f.ServerTLS; v != nil {
+		w.ServerTLS = v
+	}
+	if v := f.SessionTimeout; v != nil {
+		w.SessionTimeout = v
+	}
+	if v := f.SessionTimeout; v != nil {
+		w.SessionTimeout = v
+	}
+	if v := f.QueryTimeout; v != nil {
+		w.QueryTimeout = v
+	}
+	if v := f.BaseUserAttr; v != nil {
+		w.BaseUserAttr = v
+	}
+	if v := f.BaseDN; v != nil {
+		w.BaseDN = v
+	}
+	if v := f.UsersDN; v != nil {
+		w.UsersDN = v
+	}
+	if v := f.GroupsDN; v != nil {
+		w.GroupsDN = v
+	}
+	if v := f.ActiveAttribute; v != nil {
+		w.ActiveAttribute = v
+	}
+	if v := f.ActiveAttributeAllowedValue; v != nil {
+		w.ActiveAttributeAllowedValue = v
+	}
+	if v := f.AdminUserGroupCN; v != nil {
+		w.AdminUserGroupCN = v
+	}
+	if v := f.EditUserGroupCN; v != nil {
+		w.EditUserGroupCN = v
+	}
+	if v := f.RunUserGroupCN; v != nil {
+		w.RunUserGroupCN = v
+	}
+	if v := f.ReadUserGroupCN; v != nil {
+		w.ReadUserGroupCN = v
+	}
+	if v := f.UserApiTokenEnabled; v != nil {
+		w.UserApiTokenEnabled = v
+	}
+	if v := f.UserAPITokenDuration; v != nil {
+		w.UserAPITokenDuration = v
+	}
+	if v := f.UpstreamSyncInterval; v != nil {
+		w.UpstreamSyncInterval = v
+	}
+	if v := f.UpstreamSyncRateLimit; v != nil {
+		w.UpstreamSyncRateLimit = v
+	}
+}
+
+type WebServerLDAPSecrets struct {
+	ServerAddress     *models.SecretURL
+	ReadOnlyUserLogin *models.Secret
+	ReadOnlyUserPass  *models.Secret
```

Comment:
Is there anything we can validate up front in a `ValidateConfig` method?

Reply:
Possibly, do we do the same for DB connections for example? Connect and do a ping? Otherwise I do a test login when the LDAP module is initialized to be stored in the application struct

Reply:
I replied to this already but it didn't stick? :confused: 

Not exactly like that. We only validate the config directly at this point. I was thinking like if login and pass were required or something like that.

Reply:
Ah right I see what you mean 🙂 In this case, its a bit tougher since we expect the empty case when authentication method is not LDAP, for the above config it was easy to check because that field was present in the same struct, but for the secrets it doesn't seem like I have access to that 

```
func (w *WebServer) ValidateConfig() (err error) {
	// Validate WebServer config, assert LDAP fields when AuthMethod set to LDAP
	if *w.AuthenticationMethod == string(sessions.LDAPAuth) {
		if *w.LDAP.BaseDN == "" {
```

And secrets config 

```
type WebServerLDAPSecrets struct {
	ServerAddress     *models.SecretURL
	ReadOnlyUserLogin *models.Secret
	ReadOnlyUserPass  *models.Secret
}

func (w *WebServerLDAPSecrets) setFrom(f *WebServerLDAPSecrets) {
	if v := f.ServerAddress; v != nil {
		w.ServerAddress = v
	}
```

Is there a way I can pass or access `w.AuthenticationMethod` ?